File: Python/bioinfo/exonCalc.py
import re
import os
from  collections import OrderedDict
import time
time1 = time.time()

fileName="D:\coding\Java\CCDS.20160908.txt";

#获取染色体列表
def getChrList():
    chrs=[];
    with open(fileName,'rt') as f:
         for line in f:
             if line.startswith('#'):
                continue
             line=line.rstrip()
             lst=line.split('\t')
             if lst[-2] == '-':
                 continue
             if lst[0] not in chrs:
                 chrs.append(lst[0])
    return chrs;


#按照染色体计算独特位点个数
def calcByChr(chrName):
    lineNumber=0;#总行数
    counter=0;#行数计数器
    exonLength=0
    overlapExons=OrderedDict()
    with open(fileName,'rt') as f:
         for line in f:
             lineNumber+=1 #总行数
             if line.startswith('#'):
                continue
             line=line.rstrip()
             lst=line.split('\t')
             if lst[0] != chrName:
                 continue
             counter += 1 #进入计数的染色体总数
             if lst[-2] == '-':
                 continue
             lst[-2] = re.sub('\[|\]',' ',lst[-2])
             exons=lst[-2].split(', ')
             for exon in exons:
                 start = exon.split("-")[0] 
                 end = exon.split("-")[1]
                 for coordinate in range(int(start),int(end)+1):
                     #coordinate = lst[0] + ':'+str(i)
                     #coordinate=i
                     if coordinate not in overlapExons.keys():
                        overlapExons[coordinate] = 1
                        exonLength += 1
                        #exonLength += int(end) - int(start)
    print("chr",chrName,"共",counter,"行/",lineNumber,"行，特异位点共",exonLength,"个");
    return exonLength;



#1.获取总染色体列表
chrs=getChrList(); 
#chrs=["1"] #debug use only

#2.统计每个染色体的cds长度。累加获得总cds长度。
totalLength=0;
for chr in chrs:
    totalLength += calcByChr(chr);

print("总长度",totalLength,"bp");

time2 = time.time()
print("总耗时：",time2-time1,'s');

File: R_scripts/exonCalc.R
#rm(list=ls())

fileName="D:/coding/Java/CCDS.20160908.txt";


calcByChr=function(currentChr){
  overlapExons=c();#保存已经存在的位点
  
  #函数1：由"925941-926012"  "930154-930335" 字符串按行计算总长度
  getLengthFromArrs=function(arr){
    innerLen=0;
    for (str in arr) {
      limits <- strsplit(str,'-')[[1]]# "925941-926012"
      start=as.numeric(limits[1]);
      end=as.numeric(limits[2]);
      for (i in c(start:end)) {
        #if(is.logical(which(overlapExons==i))){print("有") }else{print("无")}
        if(length(overlapExons[overlapExons==i])==0){ #判断变量i是否在数组overlapExons中
          #print("无") 
          overlapExons <<- c(overlapExons,i);
          innerLen=innerLen+1;
        }
      }
    }
    return(innerLen)
  }
  
  
  
  lineNumber=0;#文件总行数
  counter=0;#当前染色体总行数
  totalLength=0;
  
  #=====================================
  #逐行读取文件
  con <- file(fileName, "r")
  on.exit(close(con))
  line=readLines(con,n=1) #读取第一行
  
  while( length(line) != 0 ) {
    lineNumber=lineNumber+1;
    #print(paste(lineNumber,">>",line))

    lineDice=strsplit(line,"\t")[[1]];
    chr=lineDice[1];
    
    if(chr==currentChr){
      cds=lineDice[10];
      counter=counter+1;
      if(cds!="-"){
        cds=sub('\\[','',cds);
        cds=sub('\\]','',cds); # 这个时候得到的对象还是像这样的“880073-880179, 880436-880525……”
        arrStartEnd <- strsplit(as.character(cds),', ')[[1]]# 我们先从逗号开始分割成小块
        totalLength=totalLength+getLengthFromArrs(arrStartEnd);
      }
    }
    
    if(lineNumber%%200==0) print(paste(lineNumber,",",length(overlapExons)));#进程监控
    if(lineNumber>5000) break;#性能测试
    
    line=readLines(con,n=1);#读取下一行
  }
  #close(con) #关闭文件
  #=====================================
  
  msg=paste("chr",currentChr,":",counter,"/",lineNumber,"行, 独特位点len=",totalLength);
  print(msg)
}

################################

# 如何加快R代码运行速度http://www.biotrainee.com/thread-3-6-1.html
#R里避免用for对非常大的向量历遍（小向量可以），改用apply系列的函数和避免多个数据追加等这些小细节


time1=Sys.time();
calcByChr("1")

time2=Sys.time();
print(paste("消耗时间",time2-time1,'min'))

################################

